modules/core/src/main/java/org/apache/ignite/internal/processors/service/IgniteServiceProcessor.java

```diff
@@ -0,0 +1,1764 @@
+                }
+                catch (InterruptedException ignore) {
+                    Thread.currentThread().interrupt();
+
+                    U.error(log, "Got interrupted while waiting for service to shutdown (will continue " +
+                        "stopping node): " + srvcCtx.name());
+                }
+            }
+        });
+
+        locServices.clear();
+    }
+
+    /** {@inheritDoc} */
+    @Override public void collectGridNodeData(DiscoveryDataBag dataBag) {
+        if (dataBag.commonDataCollectedFor(SERVICE_PROC.ordinal()))
+            return;
+
+        ServicesCommonDiscoveryData clusterData = new ServicesCommonDiscoveryData(
+            new ArrayList<>(registeredServices.values())
+        );
+
+        dataBag.addGridCommonData(SERVICE_PROC.ordinal(), clusterData);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onGridDataReceived(DiscoveryDataBag.GridDiscoveryData data) {
+        if (data.commonData() == null)
+            return;
+
+        ServicesCommonDiscoveryData clusterData = (ServicesCommonDiscoveryData)data.commonData();
+
+        for (ServiceInfo desc : clusterData.registeredServices())
+            registeredServices.put(desc.serviceId(), desc);
+    }
+
+    /** {@inheritDoc} */
+    @Override public void collectJoiningNodeData(DiscoveryDataBag dataBag) {
+        ArrayList<ServiceInfo> staticServicesInfo = staticallyConfiguredServices(true);
+
+        dataBag.addJoiningNodeData(SERVICE_PROC.ordinal(), new ServicesJoinNodeDiscoveryData(staticServicesInfo));
+    }
+
+    /** {@inheritDoc} */
+    @Override public void onJoiningNodeDataReceived(DiscoveryDataBag.JoiningNodeDiscoveryData data) {
+        if (data.joiningNodeData() == null)
+            return;
+
+        ServicesJoinNodeDiscoveryData joinData = (ServicesJoinNodeDiscoveryData)data.joiningNodeData();
+
+        for (ServiceInfo desc : joinData.services()) {
+            assert desc.topologySnapshot().isEmpty();
+
+            ServiceInfo oldDesc = registeredServices.get(desc.serviceId());
+
+            if (oldDesc != null) { // In case of a collision of IgniteUuid.randomUuid() (almost impossible case)
+                U.warn(log, "Failed to register service configuration received from joining node : " +
+                    "[nodeId=" + data.joiningNodeId() + ", cfgName=" + desc.name() + "]. " +
+                    "Service with the same service id already exists, cfg=" + oldDesc.configuration());
+
+                continue;
+            }
+
+            oldDesc = lookupInRegisteredServices(desc.name());
+
+            if (oldDesc == null) {
+                registeredServices.put(desc.serviceId(), desc);
+
+                continue;
+            }
+
+            if (oldDesc.configuration().equalsIgnoreNodeFilter(desc.configuration())) {
+                U.warn(log, "Ignore service configuration received from joining node : " +
```

Comment:
Log level of this message should be changed to debug. There is nothing bad in matching configurations. It may happen if some nodes have the same services in their static configuration.
The same situation with cache configurations doesn't trigger any warnings.

Reply:
Changed.